### PR TITLE
raise error if y_true not in between 0 and 1 in binary losses

### PR DESCRIPTION
**Pull request by old-school-kid**
Function `_check_ground_truth_range` return true if a value in tensor is less than 0 or more than  or NaN.  False else otherwise.
Fixes [this issue](https://github.com/keras-team/keras/issues/14046#issue-616933332)

### CODE REVIEW

**Comment by old-school-kid** [comment]
@qlzh727 would you please take a look at it?
Thank you

File: keras/losses.py [hunk 1]
```diff
@@ -1885,6 +1885,15 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
       sparse_categorical_crossentropy, from_logits=from_logits, axis=axis)
   return _ragged_tensor_apply_loss(fn, y_true, y_pred, y_pred_extra_dim=True)
 
+def _check_ground_truth_range(y_true):
+  """ Ensures ground truth labels in binary_crossentropy and 
+      binary_focal_crossentropy lie with 0.0 and 1.0
+  """
+  if tf.math.reduce_max(y_true)<= 1.0 and tf.math.reduce_min(y_true)>= 0.0 \
```
Comment (qlzh727) [inline keras/losses.py:1892]: Please use the bracket rather than \ to wrap the long line.

File: keras/losses.py [hunk 2]
```diff
@@ -1885,6 +1885,15 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
...
+    and not tf.math.is_nan(tf.math.reduce_sum(y_true)):
```
Comment (qlzh727) [inline keras/losses.py:1893]: What is this check used for?

File: keras/losses.py [hunk 3]
```diff
@@ -1885,6 +1885,15 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
...
+    return False
```
Comment (qlzh727) [inline keras/losses.py:1894]: Returning false for a passing case is quite weird. Why not just raise error in this method since the method name is `_check_ground_truth_range`

File: keras/losses.py [hunk 4]
```diff
@@ -1885,6 +1885,15 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
       sparse_categorical_crossentropy, from_logits=from_logits, axis=axis)
   return _ragged_tensor_apply_loss(fn, y_true, y_pred, y_pred_extra_dim=True)
 
+def _check_ground_truth_range(y_true):
+  """ Ensures ground truth labels in binary_crossentropy and
```
Comment (qlzh727) [inline keras/losses.py:1889]: I think the requirement for binary_crossentropy and binary_focal_crossentropy is that the label should either be 0 or 1, not just within the range of 0 and 1. eg 0.5 should be treat as a invalid label.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Comment by old-school-kid** [comment]
@qlzh727 
Please take a look now. Implemented all the changes you asked for.

File: keras/losses.py [hunk 5]
```diff
@@ -1885,6 +1885,20 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
       sparse_categorical_crossentropy, from_logits=from_logits, axis=axis)
   return _ragged_tensor_apply_loss(fn, y_true, y_pred, y_pred_extra_dim=True)
 
+def _check_ground_truth_range(y_true):
+  """ Ensures ground truth labels in binary_crossentropy and 
+      binary_focal_crossentropy are either 0.0 or 1.0.
+  """
+  y_true= tf.reshape(y_true, [-1])
+  unique, idx, count= tf.unique_with_counts(y_true)
```
Comment (qlzh727) [inline keras/losses.py:1893]: nit, please add space before the =, same for all the items below.

File: keras/losses.py [hunk 6]
```diff
@@ -1885,6 +1885,20 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
...
+  if unique.shape == 2: 
+    unique= tf.sort(unique)
+    if unique[0]== 0 and unique[1]== 1:
```
Comment (qlzh727) [inline keras/losses.py:1896]: I am not sure if this will do the correct thing when comparing the graph tensor, especially when the unique value are tf.floats

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Can you also add unit test for this?

File: keras/losses.py (hunk 4 above)
Comment (SmilingWolf) [reply keras/losses.py:1889]: What about augmentations like MixUp, where the ground truth labels are smoothed as part of the transformation?

Comment (old-school-kid) [reply keras/losses.py:1889]: @SmilingWolf ah, right!
@qlzh727 can you please confirm if the ground truth labels can be either 0 or 1 or anything in between 0 and 1?
Thank you

Comment (qlzh727) [reply keras/losses.py:1889]: I think you are referring to label_smoothing here. note that the check is applied to the raw label input which should be either 0 or 1. The y_true after label smoothing can be [0, 1]

Comment (SmilingWolf) [reply keras/losses.py:1889]: I'm not. In MixUp, the training sample is the weighted sum of two images, and the ground truth labels are themselves the weighted sums of the two original images' labels.

So, say I have image1 and and image2 with labels1 = [0, 1, 0, 1] and labels2 = [0, 1, 1, 0].
After mixing those up with a factor of 0.2, the training sample is going to be
X = (image1 * 0.2) + (image2 * (1.0 - 0.2))
and its labels are going to be
y = (labels1 * 0.2) + (labels2 * (1.0 - 0.2)) = [0.0, 1.0 , 0.8, 0.2]

This all happens within the input pipeline. For example: https://keras.io/examples/vision/mixup/#define-the-mixup-technique-function

Comment (qlzh727) [reply keras/losses.py:1889]: I see, good to know. In that sense, we might just want to limit the value to range [0, 1], and leave a comment about why we do this. The normal use case (the one leave in the docstring) is probably the one ppl will use.

**Comment by old-school-kid** [comment]
@qlzh727 
Considering @SmilingWolf comments I again went back to check max and min. Do I have to add unit test for this?
Thank you

File: keras/losses.py [hunk 7]
```diff
@@ -1885,6 +1885,12 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
       sparse_categorical_crossentropy, from_logits=from_logits, axis=axis)
   return _ragged_tensor_apply_loss(fn, y_true, y_pred, y_pred_extra_dim=True)
 
+def _check_ground_truth_range(y_true):
+  """ Ensures ground truth labels in binary_crossentropy and 
+      binary_focal_crossentropy are between 0.0 and 1.0.
+  """
+  if tf.math.reduce_max(y_true)> 1.0 or tf.math.reduce_min(y_true)< 0.0:
```
Comment (qlzh727) [inline keras/losses.py:1892]: Please add space before and after ">" and "<"

**Comment by qlzh727** [comment]
> @qlzh727 Considering @SmilingWolf comments I again went back to check max and min. Do I have to add unit test for this? Thank you

Thanks for the update. Please add unit test for this.

**Comment by old-school-kid** [comment]
@qlzh727 
Updated the requested changes. Do I have to make any changes in docstring anywhere or any other changes in general?

**Comment by qlzh727** [comment]
> @qlzh727 Updated the requested changes. Do I have to make any changes in docstring anywhere or any other changes in general?

Looks good. Please update the docstring about the new error will be raise if the y_true is not in the valid range.

**Comment by old-school-kid** [comment]
@qlzh727 please take a look. Thanks

File: keras/losses.py [hunk 8]
```diff
@@ -502,7 +502,7 @@ class BinaryCrossentropy(LossFunctionWrapper):
   Use this cross-entropy loss for binary (0 or 1) classification applications.
   The loss function requires the following inputs:
 
-  - `y_true` (true label): This is either 0 or 1.
+  - `y_true` (true label): This is either 0 or 1. Value
```
Comment (qlzh727) [inline keras/losses.py:505]: Please remove the tailing `Value` here.

**Review by qlzh727 — APPROVED** [review]

**Comment by old-school-kid** [comment]
@qlzh727 would you please take a look at it again?

**Comment by old-school-kid** [comment]
@gbaned could you please add the `ready to pull` ticket since @qlzh727 approved the changes. Thank you

**Comment by qlzh727** [comment]
Sorry for the long wait. We are running some internal global tests to make sure it doesn't break any existing tests.

File: keras/losses.py [hunk 9]
```diff
@@ -1885,6 +1891,12 @@ def _ragged_tensor_sparse_categorical_crossentropy(y_true,
       sparse_categorical_crossentropy, from_logits=from_logits, axis=axis)
   return _ragged_tensor_apply_loss(fn, y_true, y_pred, y_pred_extra_dim=True)
 
+def _check_ground_truth_range(y_true):
+  """ Ensures ground truth labels in binary_crossentropy and 
+      binary_focal_crossentropy are between 0.0 and 1.0.
+  """
+  if tf.math.reduce_max(y_true) > 1.0 or tf.math.reduce_min(y_true) < 0.0:
```
Comment (fchollet) [inline keras/losses.py:1898]: You can't do value checks like this, because that will only work in eager mode, not in graph mode. In general value checks should be avoided because they add computation overhead. But if they're absolutely necessary, they should be implemented via a symbolic op, such as `assert_less` -- https://www.tensorflow.org/api_docs/python/tf/debugging/assert_less

In this case, it does not seem like a good idea to add this check because of the overhead. If we do it, we should restrict it to only run in eager execution (by checking https://www.tensorflow.org/api_docs/python/tf/executing_eagerly)

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. Please restrict the check to eager execution (via https://www.tensorflow.org/api_docs/python/tf/executing_eagerly) and please remove the docstrings modifications.

File: keras/losses.py (hunk 9 above)
Comment (old-school-kid) [reply keras/losses.py:1898]: Can we just drop the PR, since it's understood that binary losses are supposed to be for 0's and 1's. Using for values not in range [0,1] makes no sense.

**Comment by fchollet** [comment]
> Can we just drop the PR, since it's understood that binary losses are supposed to be for 0's and 1's. Using for values not in range [0,1] makes no sense.

Sounds good, thank you for the effort.

**Comment by Suzan009** [comment]
@old-school-kid @qlzh727 
In this competition, the task is a regression (1,2,3,...100). But practitioners found it that, it was useful to treat this task as a classification problem. And so they normalized the target i.e (0.1, 0.2, 0.3, ..1.0) and use sigmoid/binary_cross entropy.  More information. https://www.kaggle.com/c/petfinder-pawpularity-score/discussion/275094

In this PR, does it disable the opportunity to use binary cross-entropy as a loss function in the above cases? All the tensorflow [notebooks](https://www.kaggle.com/c/petfinder-pawpularity-score/code) will break down for these changes?

**Comment by old-school-kid** [comment]
@Suzan009 no this feature checks if values are in range [0, 1]. So no, your task would have run successfully if this PR was merged. Oh also this PR has not been merged as it would have added overheads. So I dont think you will face a problem.
Hope this helps.
